src/test/java/com/google/api/generator/engine/JavaCodeGeneratorTest.java

```diff
@@ -0,0 +1,885 @@
+import java.util.Stack;
+import org.junit.Test;
+
+public class JavaCodeGeneratorTest {
+  // Create shared VaporReferences.
+  private static final VaporReference stubRef = createVaporReference("com.google.gax.grpc", "Stub");
+  private static final VaporReference bookKindRef =
+      createVaporReference("com.google.exmaple.library.v1", "BookKind");
+  private static final VaporReference libraryServiceRef =
+      createVaporReference("com.google.exmaple.library", "LibraryService");
+  private static final VaporReference shelfClassRef =
+      createVaporReference("com.google.example.library.core", "Shelf");
+  private static final VaporReference bookClassRef =
+      createVaporReference("com.google.example.library.core", "Book");
+  private static final VaporReference novelClassRef =
+      createVaporReference("com.google.example.library.core", "Novel");
+  // Create shared ConcreteReferences.
+  private static final ConcreteReference shelfListRef =
+      ConcreteReference.builder()
+          .setClazz(List.class)
+          .setGenerics(Arrays.asList(shelfClassRef))
+          .build();
+  private static final ConcreteReference stringRef = ConcreteReference.withClazz(String.class);
+  private static final ConcreteReference shelfMapRef =
+      ConcreteReference.builder()
+          .setClazz(HashMap.class)
+          .setGenerics(Arrays.asList(stringRef, shelfClassRef))
+          .build();
+  // Create shared variables.
+  private static final Variable shelfVar = createVarFromVaporRef(shelfClassRef, "shelf");
+  private static final Variable shelfListVar = createVarFromConcreteRef(shelfListRef, "shelfList");
+  private static final Variable shelfMapVar = createVarFromConcreteRef(shelfMapRef, "shelfMap");
+  private static final Variable shelfNameVar = createVarFromType(TypeNode.STRING, "shelfName");
+  private static final Variable seriesNumVar = createVarFromType(TypeNode.INT, "seriesNum");
+  private static final Variable shelfServiceNameVar =
+      createVarFromType(TypeNode.STRING, "shelfServiceName");
+  private static final Variable bookKindVar = createVarFromVaporRef(bookKindRef, "bookKind");
+
+  @Test
+  public void validJavaClass() {
+    // Create outer class variableDecls.
+    // [code] private static final String serviceName = "LibraryServiceStub";
+    VariableExpr serviceName = createServiceNameVarExpr();
+    AssignmentExpr serviceNameDel =
+        createAssignmentExpr(
+            serviceName, ValueExpr.withValue(StringObjectValue.withValue("LibraryServiceStub")));
+
+    // [code] protected List<Shelf> shelfList;
+    VariableExpr shelfListExpr = createShelfListVarExpr();
+
+    // [code] public static HashMap<String, Shelf> shelfMap;
+    VariableExpr shelfMapExpr = createShelfMapVarExpr();
+
+    // Create the LibraryServiceStub constructor
+    MethodDefinition libraryServiceCtor = createLibServiceCtor();
+
+    // Create nested class Shelf.
+    VariableExpr shelfNameDel = createVarPublicDeclExpr(shelfNameVar);
```

Comment:
Consider also breaking out the nested classes into separate methods.

Reply:
Sure, moved the nested classes to separate methods, only left the outer class definition in the main body. Thanks!